crates/wasmtime/src/values.rs

```diff
@@ -187,22 +199,44 @@ impl Val {
     pub(crate) fn into_table_element(
         self,
         store: &mut StoreOpaque,
-        ty: ValType,
+        ty: RefType,
     ) -> Result<TableElement> {
         match (self, ty) {
-            (Val::FuncRef(Some(f)), ValType::FuncRef) => {
+            (
+                Val::FuncRef(Some(f)),
+                RefType {
+                    heap_type: HeapType::Func,
+                    ..
+                },
+            ) => {
                 if !f.comes_from_same_store(store) {
                     bail!("cross-`Store` values are not supported in tables");
                 }
                 Ok(TableElement::FuncRef(
                     f.caller_checked_anyfunc(store).as_ptr(),
                 ))
             }
-            (Val::FuncRef(None), ValType::FuncRef) => Ok(TableElement::FuncRef(ptr::null_mut())),
-            (Val::ExternRef(Some(x)), ValType::ExternRef) => {
-                Ok(TableElement::ExternRef(Some(x.inner)))
-            }
-            (Val::ExternRef(None), ValType::ExternRef) => Ok(TableElement::ExternRef(None)),
+            (
+                Val::FuncRef(None),
+                RefType {
+                    heap_type: HeapType::Func,
+                    ..
+                },
+            ) => Ok(TableElement::FuncRef(ptr::null_mut())),
+            (
+                Val::ExternRef(Some(x)),
+                RefType {
+                    heap_type: HeapType::Extern,
+                    ..
+                },
+            ) => Ok(TableElement::ExternRef(Some(x.inner))),
+            (
+                Val::ExternRef(None),
+                RefType {
+                    heap_type: HeapType::Extern,
+                    ..
+                },
+            ) => Ok(TableElement::ExternRef(None)),
             _ => bail!("value does not match table element type"),
```

Comment:
I think that this `match` is going to need to grow more cases or otherwise embedders can't interact with typed function tables in the embedder API. That's ok for a first pass and can be deferred to a future PR but it's functionality we'll eventually want to fill out.

Reply:
To bump this, I think it would be good to have at least have a specific error for `HeapType::Index` which indicates that support is missing.

Reply:
Added, thanks!